### PR TITLE
Add to @Test interface fields IDs and issues

### LINKED ISSUE #2003: [improvement] Add to @Test interface fields IDs and issues
As an auto tester, I want to be able to associate an autotest with a test case in the test management system and associate the test with an issue in the task tracker system.

For autotest management, it is worth adding auxiliary fields to the `@Test` interface so that the end user can manage the presentation of the results in the checklists.

Fields:
* `String[] ids = default {};` - test case IDs. For example in [testrail](https://www.gurock.com/testrail), [kanoah](https://kanoah.com/) or another test management system.
* `String[] issues = default {};` - features or/and bugs in the task tracker system. For example in jira, mantis, YouTrack, etc.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2003

Fixes #2003 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

**Review by juherr — COMMENTED** [review]
Some few remarks.

File: src/main/java/org/testng/annotations/CustomAttributes.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * Represents a means to add in custom attributes to @{@link Test} annotated tests.
+ */
+@Retention(java.lang.annotation.RetentionPolicy.RUNTIME)
+@Target({METHOD, TYPE})
+public @interface CustomAttributes {
+
+  /**
+   * @return - The key for the custom attribute
+   */
+  String name();
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/annotations/CustomAttributes.java]: fixed.

File: src/test/java/test/custom/CustomAttributesListener.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+import org.testng.IInvokedMethod;
+import org.testng.IInvokedMethodListener;
+import org.testng.ITestResult;
+import org.testng.annotations.CustomAttributes;
+import org.testng.collections.Lists;
+
+public class CustomAttributesListener implements IInvokedMethodListener {
+
+  private List<CustomAttributes> attributes = Lists.newArrayList();
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/test/java/test/custom/CustomAttributesListener.java]: Fixed (Oh btw this is test code)

File: src/test/java/test/custom/CustomAttributesListener.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+  @Override
+  public void afterInvocation(IInvokedMethod method, ITestResult testResult) {
+    this.attributes.addAll(Arrays.asList(method.getTestMethod().getAttributes()));
+  }
+
+  public List<CustomAttributes> getAttributes() {
+    return attributes;
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/test/java/test/custom/CustomAttributesListener.java]: Fixed (Oh btw this is test code)

File: src/test/java/test/custom/CustomAttributesTransformer.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    };
+    annotation.setAttributes(attributes);
+  }
+
+  public static class MoreAttributes implements CustomAttributes {
+
+    private String key;
+    private String[] values;
+
+    public MoreAttributes(String key, String[] values) {
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/test/java/test/custom/CustomAttributesTransformer.java]: Fixed.

File: src/main/java/org/testng/annotations/Test.java [hunk 5]
```diff
@@ -143,4 +143,10 @@
 
   /** The scheduling priority. Lower priorities will be scheduled first. */
   int priority() default 0;
+
+  /**
+   * @return - An array of {@link CustomAttributes} that represents a set of custom attributes for a
+   * test method.
+   */
+  CustomAttributes[] attributes() default {};
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/annotations/Test.java]: Agreed. Renamed.

**Comment by krmahadevan** [comment]
@juherr - I amended the commit and pushed back the changes. Please help take a look

**Review by juherr — APPROVED** [review]
You can merge when you want. Just check my last comment.

File: src/main/java/org/testng/ITestNGMethod.java [hunk 6]
```diff
@@ -247,4 +248,12 @@ default boolean isDataDriven() {
   default IParameterInfo getFactoryMethodParamsInfo() {
     return null;
   }
+
+  /**
+   * @return - An array of {@link CustomAttribute} that represents the custom attributes associated
+   * with a test.
+   */
+  default CustomAttribute[] getAttributes() {
+    return null;
```
[the remark this answers is not in the document]
Comment (krmahadevan) [reply src/main/java/org/testng/ITestNGMethod.java]: fixed

**Pull request merged**